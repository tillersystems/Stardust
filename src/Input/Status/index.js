import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { Icon } from '../..';
import Theme from '../../Theme';

/** Lookup object for icon name from status. */
const statusToIconName = {
  loading: 'circle-notch',
  info: 'info',
  success: 'check-mark',
  warning: 'exclamation-mark',
  error: 'cross',
  search: 'search',
};

/**
 * Renders a status badge for an input.
 *
 * @param {string} status - The status to display.
 *
 * @return {jsx}
 */
const Status = ({ className, status }) => (
  <div className={className}>
    <Icon
      color={status === 'loading' ? Theme.palette.primary.default : Theme.palette.white}
      name={statusToIconName[status]}
      width={['loading', 'info', 'warning'].includes(status) ? '2rem' : '1.1rem'}
      height={['loading', 'info', 'warning'].includes(status) ? '2rem' : '1.1rem'}
      spin={status === 'loading'}
    />
  </div>
);

/** Display name. */
Status.displayName = 'Input.Status';

/** Props types. */
const { string } = PropTypes;
Status.propTypes = {
  className: string,
  status: string.isRequired,
};

/** Default props. */
Status.defaultProps = {
  className: '',
};

export default styled(Status)`
  display: flex;
  align-items: center;
  justify-content: center;

  width: 1.8rem;
  height: 1.8rem;

  margin-right: ${({ theme: { dimensions } }) => dimensions.small};

  border-radius: 50%;

  ${({ status }) =>
    status === 'loading' &&
    css`
      background: ${({ theme: { palette } }) => palette.white};
    `};

  ${({ status }) =>
    status === 'info' &&
    css`
      background: ${({ theme: { palette } }) => palette.primary.default};
    `};

  ${({ status }) =>
    status === 'success' &&
    css`
      background: ${({ theme: { palette } }) => palette.success.default};
    `};

  ${({ status }) =>
    status === 'warning' &&
    css`
      background: ${({ theme: { palette } }) => palette.warning.default};
    `};

  ${({ status }) =>
    status === 'error' &&
    css`
      background: ${({ theme: { palette } }) => palette.failure.default};
    `};

  ${({ status }) =>
    status === 'search' &&
    css`
      background: ${({ theme: { palette } }) => palette.lightGrey};
    `};

  ${({ status, hasFocus }) =>
    status === 'search' &&
    hasFocus &&
    css`
      background: ${({ theme: { palette } }) => palette.primary.default};
    `};
`;
