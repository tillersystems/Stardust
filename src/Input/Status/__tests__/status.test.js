import React from 'react';

import Status from '..';

describe('<Status />', () => {
  test('should render without a problem', () => {
    const { container } = render(<Status status="info" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should have different width and height on icon', () => {
    const { container } = render(<Status status="search" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should have different color on icon', () => {
    const { container } = render(<Status status="loading" />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
