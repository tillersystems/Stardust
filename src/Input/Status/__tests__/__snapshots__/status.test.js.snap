// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Status /> should have different color on icon 1`] = `
.c1 {
  display: inline-block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.8rem;
  height: 1.8rem;
  margin-right: 0.7rem;
  border-radius: 50%;
  background: hsl(0,100%,100%);
}

<div
  class="c0"
  data-testid="status"
>
  <svg
    aria-hidden="true"
    class="c1"
    focusable="false"
    height="2rem"
    role="img"
    viewBox="0 0 512 512"
    width="2rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <path
        d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
        data-testid="icon-svg-path"
        fill="hsl(200,74%,46%)"
        fill-rule="evenodd"
      />
    </g>
  </svg>
</div>
`;

exports[`<Status /> should have different width and height on icon 1`] = `
.c1 {
  display: inline-block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.8rem;
  height: 1.8rem;
  margin-right: 0.7rem;
  border-radius: 50%;
}

<div
  class="c0"
  data-testid="status"
>
  <svg
    aria-hidden="true"
    class="c1"
    focusable="false"
    height="1.33rem"
    role="img"
    viewBox="0 0 512 512"
    width="1.33rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <path
        d="M384.68 354.51l78.405 78.405c8.331 8.331 8.331 21.839 0 30.17-8.331 8.331-21.839 8.331-30.17 0l-78.404-78.404c-32.847 26.274-74.51 41.986-119.844 41.986-106.039 0-192-85.962-192-192 0-106.039 85.961-192 192-192 106.038 0 192 85.961 192 192 0 45.333-15.712 86.997-41.986 119.844zm-42.464-16.24C368.09 311.417 384 274.9 384 234.667c0-82.475-66.859-149.334-149.333-149.334-82.475 0-149.334 66.86-149.334 149.334S152.193 384 234.667 384c40.233 0 76.75-15.91 103.603-41.784a21.534 21.534 0 0 1 3.946-3.946z"
        data-testid="icon-svg-path"
        fill="hsl(206,16%,60%)"
        fill-rule="evenodd"
      />
    </g>
  </svg>
</div>
`;

exports[`<Status /> should render without a problem 1`] = `
.c1 {
  display: inline-block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 1.8rem;
  height: 1.8rem;
  margin-right: 0.7rem;
  border-radius: 50%;
  background: hsl(200,74%,46%);
}

<div
  class="c0"
  data-testid="status"
>
  <svg
    aria-hidden="true"
    class="c1"
    focusable="false"
    height="2rem"
    role="img"
    viewBox="0 0 512 512"
    width="2rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <path
        d="M256 492C125.66 492 20 386.34 20 256S125.66 20 256 20s236 105.66 236 236-105.66 236-236 236zm21.333-150.667V256c0-11.782-9.55-21.333-21.333-21.333-11.782 0-21.333 9.55-21.333 21.333v85.333c0 11.782 9.55 21.334 21.333 21.334 11.782 0 21.333-9.552 21.333-21.334zM256 192c11.782 0 21.333-9.551 21.333-21.333 0-11.782-9.55-21.334-21.333-21.334-11.782 0-21.333 9.552-21.333 21.334S244.217 192 256 192z"
        data-testid="icon-svg-path"
        fill="hsl(0,100%,100%)"
        fill-rule="evenodd"
      />
    </g>
  </svg>
</div>
`;
